### PR TITLE
Devito smooth

### PR DESCRIPTION
**Pull request by rhodrin**
This PR:

1. Adds some additional functionality to the `gaussian_smooth` builtin. It should now work correctly with arrays of floats and now accepts `tuple`'s of `sigma`. Further tests relating to these functionalities have been added.
2. ~~Replaces `scipy_smooth` with `devito_smooth`~~ Replaces `scipy_smooth` with a method in `Model` calling the `gaussian_smooth` builtin.

### CODE REVIEW

File: devito/builtins.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def smooth(f, g, axis=None):
         dv.Operator(dv.Eq(f, g.avg(dims=axis)), name='smoother')()
 
 
-def gaussian_smooth(f, sigma=1, _order=4, mode='reflect'):
+def gaussian_smooth(f, sigma=1, _truncate=4.0, mode='reflect'):
```
Comment (FabioLuporini) [inline devito/builtins.py]: this needs a proper docstring? At least the Parameters

also, why `_truncate` and not `truncate` ?

File: devito/builtins.py [hunk 2]
```diff
@@ -123,46 +135,61 @@ def fset(f, g):
         else:
             raise NotImplementedError
 
-    lw = int(_order*sigma + 0.5)
+    try:
+        dtype = f.dtype.type
```
Comment (FabioLuporini) [inline devito/builtins.py]: ? this is for numpy arrays ? if so, maybe leave a comment?

File: devito/builtins.py [hunk 3]
```diff
@@ -123,46 +135,61 @@ def fset(f, g):
...
+    except AttributeError:
+        dtype = f.dtype
+
+    # TODO: Add s = 0 dim skip option
```
Comment (FabioLuporini) [inline devito/builtins.py]: leftover?

File: devito/builtins.py [hunk 4]
```diff
@@ -123,46 +135,61 @@ def fset(f, g):
...
+    lw = tuple([int(_truncate*float(s) + 0.5) for s in as_tuple(sigma)])
```
Comment (FabioLuporini) [inline devito/builtins.py]: nitpicking, pythonicity:
no need for the square brackets 
`int(_truncate*float(s) + 0.5) for s in as_tuple(sigma)` creates a generator, and `tuple(...)` expands it.

File: devito/builtins.py [hunk 5]
```diff
@@ -123,46 +135,61 @@ def fset(f, g):
...
+
+    if len(lw) == 1 and len(lw) < len(f.shape):
+        lw = len(f.shape)*(lw[0], )
+        sigma = len(f.shape)*(as_tuple(sigma)[0], )
+    elif len(lw) == len(f.shape):
+        sigma = as_tuple(sigma)
+    else:
+        raise ValueError("sigma must be an integer or a tuple of length" +
```
Comment (FabioLuporini) [inline devito/builtins.py]: nitpicking: throughout the codebase, we use backtips around keywords in error messages

\`signa\` must be an integer or a tuple of length \`f.ndim\`

File: devito/builtins.py [hunk 6]
```diff
@@ -123,46 +135,61 @@ def fset(f, g):
...
+                         " function.dimensions.")
 
     # Create the padded grid:
     objective_domain = ObjectiveDomain(lw)
+    # NOTE: This 'try' is required for distributed `f`.
```
Comment (FabioLuporini) [inline devito/builtins.py]: what happens if you use `shape_global` in both cases?

File: devito/builtins.py (hunk 1 above)
Comment (rhodrin) [reply devito/builtins.py]: There was a reason, but I don't recall what it was and is now probably irrelevant. Will change.

File: devito/builtins.py (hunk 2 above)
Comment (rhodrin) [reply devito/builtins.py]: Yes. Will add comment.

File: devito/builtins.py (hunk 3 above)
Comment (rhodrin) [reply devito/builtins.py]: No, still need to add that functionality.

File: devito/builtins.py (hunk 4 above)
Comment (rhodrin) [reply devito/builtins.py]: Ta.

File: devito/builtins.py (hunk 5 above)
Comment (rhodrin) [reply devito/builtins.py]: I will remember to start writing these properly some day...

File: devito/builtins.py (hunk 6 above)
Comment (rhodrin) [reply devito/builtins.py]: Input can be a numpy array so won't have that property?

File: devito/builtins.py [hunk 7]
```diff
@@ -111,58 +123,86 @@ def __init__(self, lw):
...
+
         lhs.append(f_c)
         rhs.append(f_o)
         options.append({'subdomain': grid.subdomains['objective_domain']})
 
         mapper[d] = {'lhs': lhs, 'rhs': rhs, 'options': options}
 
-    initialize_function(f_c, f.data[:], lw,
+    initialize_function(f_c, f, lw,
                         mapper=mapper,
```
Comment (FabioLuporini) [inline devito/builtins.py]: nitpicking: three lines -> one line ?

File: devito/builtins.py [hunk 8]
```diff
@@ -252,23 +301,23 @@ def initialize_function(function, data, nbl, mapper=None, mode='constant',
...
 
         b = [min(l) for l in (w for w in (buff(i, j) for i, j in zip(local_size, halo)))]
         if any(np.array(b) < 0):
             raise ValueError("Function `%s` halo is not sufficiently thick." % function)
 
-    for d in function.space_dimensions:
+    for d, n in zip(function.space_dimensions, as_tuple(nbl)):
         dim_l = dv.SubDimension.left(name='abc_%s_l' % d.name, parent=d,
-                                     thickness=nbl)
+                                     thickness=n)
```
Comment (FabioLuporini) [inline devito/builtins.py]: nitpicking : chance for inlining ?

File: examples/seismic/model.py [hunk 9]
```diff
@@ -686,6 +691,24 @@ def vp(self, vp):
...
+        ----------
+        physical_parameters : string or tuple of string
+            Names of the fields to be smoothed.
+        sigma : float
+            Standard deviation of the smoothing operator.
+        """
+        model_parameters = self.physical_params()
+        for i in physical_parameters:
+            # NOTE/FIXME: It would seem we can just remove devito_smooth now
+            # and call the gaussian smooth builtin directly here?
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: agreed

File: examples/seismic/model.py [hunk 10]
```diff
@@ -231,16 +231,21 @@ def demo_model(preset, **kwargs):
...
+        theta = .5*(v - 1.5)
         phi = None
         if len(shape) > 2:
-            phi = scipy_smooth(.25*(v - 1.5), shape)
+            phi = .25*(v - 1.5)
+        model = Model(space_order=space_order, vp=v, origin=origin, shape=shape,
+                      dtype=dtype, spacing=spacing, nbl=nbl, epsilon=epsilon,
+                      delta=delta, theta=theta, phi=phi, **kwargs)
+        if len(shape) > 2:
+            model.smooth(('epsilon', 'delta', 'theta', 'phi'))
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: This should be a Model constructor argument

File: devito/builtins.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/builtins.py]: ah, I see... perhaps you can add `shape = ...` **inside** the try-except that you already use for `dtype`

or at this point, you may well use `if isinstance(x, np.ndarray) ...`

anyway, nitpicking again, sorry

File: examples/seismic/model.py (hunk 9 above)
Comment (rhodrin) [reply examples/seismic/model.py]: Good. Shall I also remove `scipy_smooth`?

Comment (FabioLuporini) [reply examples/seismic/model.py]: I think so

**Pull request merged**